### PR TITLE
Modular MiniGame system (Abandoned Crates)

### PR DESCRIPTION
**Pull request by MaxIsJoe**
## Purpose

Jack has proposed an idea for me that caught me interest and while working on it, I've decided to make a modular mini-game system that can be re-used anywhere. The idea of this is that mini-games should not just be tied to crates and we can add all sorts of mini-games in the future without any restrictions.

## Changelog:

CL: [New] - Added goon-style abandoned crates that can be unlocked by playing minigames.
CL: [New] - Added a "voice-activated guess-the-number" minigame.

### CODE REVIEW

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs [hunk 1]
```diff
@@ -0,0 +1,34 @@
+using UnityEngine;
+
+namespace MiniGames
+{
+	/// <summary>
+	/// MiniGameModules are gameObjects that sit on the gameObject that has the IMiniGame Interface.
+	/// They are meant to be used when an object can house more than one minigame or if the minigame's behavior is too complex
+	/// to be stored in just the StartMiniGame() method.
+	/// </summary>
+	public class MiniGameModule : MonoBehaviour
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs]: make abstract

File: UnityProject/Assets/Scripts/Objects/Closets/ClosetControl.cs [hunk 2]
```diff
@@ -5,14 +5,18 @@
...
+using UnityEngine.Events;
+using Random = UnityEngine.Random;
 
 namespace Objects
 {
 	/// <summary>
 	/// Allows closet to be opened / closed / locked
 	/// </summary>
-	public class ClosetControl : NetworkBehaviour, IServerSpawn, ICheckedInteractable<PositionalHandApply>, IRightClickable, IExaminable, IEscapable
+	public class ClosetControl : NetworkBehaviour, IServerSpawn, ICheckedInteractable<PositionalHandApply>, IRightClickable, IExaminable, IEscapable, IMiniGame
```
Comment (corp-0) [inline UnityProject/Assets/Scripts/Objects/Closets/ClosetControl.cs]: I would rather make a new component with the minigame that uses ``ClosetControl``. This minigame also won't be used in most of crates/lockers, so it is better to keep the responsibility of this class scoped. The abandoned crates would be their own prefab which add this component and the minigame component.

TLDR: invert the references so a component that implements ``IMinigame`` uses ``ClosetControl`` and not the other way.

**Review by corp-0 — CHANGES_REQUESTED** [review]

File: UnityProject/Assets/Scripts/Objects/Closets/ClosetControl.cs (hunk 2 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/ClosetControl.cs]: On it when I jump on rider 👍

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs [hunk 3]
```diff
@@ -0,0 +1,87 @@
+using System.Collections.Generic;
+using MiniGames;
+using Mirror;
+using UnityEngine;
+
+namespace Objects.Closets
+{
+	public class AbandonedCrates : ClosetControl, IMiniGame
```
Comment (NoooneyDude) [inline UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: It doesn't make much sense for a minigame to _be_ a closet. ~~Rather, a minigame can _have_ a closet~~ I see, we can attach any minigame to this closet. But the minigame delegator shouldn't extend closets.

`GetComponent<ClosetControl>.DoThingsToClosetControl()` would be more appropriate. ~~This would also allow this class to inherit from `MiniGameModule`, if that makes more sense to you.~~

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: The minigame is not the closet, AbandonedCrate is a closet that has it's own behavior.

Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: All of this code is about minigames, not a closet.

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: Like I said, Abandoned Crates are closets with their own behavior. Their behavior is that they use minigames to be opened which in this context makes sense because we want the closets' behavior for storing objects and closing/opening them. The minigames take place the open/close interactions and that's the whole point of those crates, minigames.

Am I misreading something?

Comment (NoooneyDude) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: The problem is this can't be easily used for any type of `ClosetControl`. Sure, we might only have one `ClosetControl` script for now, but if we add others, then we could no longer use this script and we'd need to refactor it, if we wanted to add minigames to them. However, if we use _composition_, we don't need to worry about what the `ClosetControl` looks like, just that we trust it to do its job.

So, I am indicating a design issue. But I won't fight you on it.

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: You actually make a really good point, but:
The problem with separating this into two scripts is that Unity has a problem with component hierarchies, which have caused me headaches multiples times before while working on stuff that involves interactions. So, while I do agree with you, I don't want this to randomly break in the future because Unity decided to shuffle component positions around.

Comment (corp-0) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: I still believe this could very well just be its own component and use a reference to Closet control. Don't see why it should extend it.

Comment (Alainx277) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: I agree with Gilles here.

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+			{
+				Logger.LogError("[ClosetControl/MiniGames] - Found MiniGame tracker but no minigames found!");
+				return;
+			}
+			currentMiniGameIndex = Random.Range(0, miniGameModules.Count - 1);
+			miniGameModules[currentMiniGameIndex].Setup(miniGameTracker, gameObject);
+			SetLock(Lock.Locked);
+		}
+
+		protected override void InteractionChecks(PositionalHandApply interaction)
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: I feel like it would make more sense to override `TryToggleLock`? I'd prefer an event approach, but overriding a more specific method is also fine.

File: UnityProject/Assets/Scripts/MiniGames/IMiniGame.cs [hunk 5]
```diff
@@ -0,0 +1,9 @@
+namespace MiniGames
+{
+	public interface IMiniGame
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/MiniGames/IMiniGame.cs]: I don't see this interface used anywhere?

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs [hunk 6]
```diff
@@ -0,0 +1,35 @@
...
+		/// Use the Setup() function only when you're assigned events programatically and not from the inspector.
+		/// </summary>
+		public virtual void Setup(MiniGameResultTracker tracker, GameObject parent)
+		{
+			Tracker = tracker;
+			MiniGameParent = parent;
+			Tracker.OnStartGame.AddListener(StartMiniGame);
+		}
+
+		public virtual void StartMiniGame()
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs]: If there's no default implementation it should be abstract.

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs [hunk 7]
```diff
@@ -0,0 +1,35 @@
...
+		{
+			//Add your game setup/reset/ui show/etc here
+		}
+
+		protected virtual void OnGameDone(bool t)
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs:35]: This doesn't seem to be used in the context of the base class. Can you remove it from here?

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs (hunk 3 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: I feel like almost everything here belongs into the minigame tracker.

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+using UnityEngine;
+
+namespace Objects.Closets
+{
+	public class AbandonedCrates : ClosetControl, IMiniGame
+	{
+
+		[SerializeField] private MiniGameResultTracker miniGameTracker;
+		[SerializeField] private List<MiniGameModule> miniGameModules;
+		[SyncVar] private int currentMiniGameIndex = -1;
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: Why is this synced to the client?

**Review by Alainx277 — CHANGES_REQUESTED** [review]

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs (hunk 6 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs]: abstracc

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs (hunk 8 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: bi-product of old code while testing, removing.

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs (hunk 7 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs:35]: it is being used

File: UnityProject/Assets/Scripts/MiniGames/IMiniGame.cs (hunk 5 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/MiniGames/IMiniGame.cs]: It is being used.

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs (hunk 4 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: Why would you "Try" to do something that should be a forced state?
This feels like it's going to result in bugs, so I'm not taking risks.

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs (hunk 3 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: can you explain to me how I can move this to the minigame tracker? I don't see how It makes sense to be moved there.

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: also, again, read what I said about interactions. How do you want to trigger the minigame if you guys didn't like me putting it inside the closet code and I have told you already how having two interactable interfaces causes one to be priortized over the other.

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs (hunk 7 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs:35]: It's only being used in the child class, never in the context of `MiniGameModule`. So why have it on the base class?

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs (hunk 3 above)
Comment (Alainx277) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: I thought a minigame was supposed to be generic enough to be used for multiple things. I reckon the minigame tracker (maybe call it a manager) would receive requests for a minigame, start the minigame and handle the minigame success/fail.
Why does the abandoned crate check for multiple minigame modules? It should not need to care about that.
About the interaction: We want to react when a player tries to open the locked crate. Why not create an event that fires if it's locked, which is listened to by a minigame tracker, which starts a random minigame. That way the crate code doesn't care about minigames (as it should be imo).

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs]: Moved out the functionality out of the closet control code

However, expect this to randomly break sometime in the future when Unity decides to scramble the position of components again.

**Comment by MaxIsJoe** [comment]
this should be ready for review again

**Comment by MaxIsJoe** [comment]
Fixed some codacy issues, ignore the ones related to events because it thinks they're not being used but they are in the inspector.

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+			Tracker = tracker;
+			MiniGameParent = parent;
+			Tracker.OnStartGame.AddListener(StartMiniGame);
+		}
+
+		//Add your game setup/reset/ui show/etc here
+		public abstract void StartMiniGame();
+
+		//Add your logic here when a player has won or lost a game or has left the game
+		protected abstract void OnGameDone(bool t);
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs]: Would it make sense to have this call `Tracker.OnGameEnd` so it's not forgotten?

File: UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs [hunk 10]
```diff
@@ -0,0 +1,78 @@
...
+using UnityEngine;
+using Random = UnityEngine.Random;
+
+namespace Objects.Closets
+{
+	public class AbandonedCrates : MonoBehaviour, IServerSpawn, ICheckedInteractable<HandApply>
+	{
+
+		[SerializeField] private MiniGameResultTracker miniGameTracker;
+		[SerializeField] private List<MiniGameModule> miniGameModules;
```
Comment (Alainx277) [inline UnityProject/Assets/Scripts/Objects/Closets/AbandonedCrates.cs:12]: I see that the crate is still linked directly to minigames. It's fine for now, just keep in mind that everything that has a minigame will need to do this too.

**Review by Alainx277 — CHANGES_REQUESTED** [review]
Looks much better. Just a small nit.

File: UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs (hunk 9 above)
Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs]: what do you mean

Comment (corp-0) [reply UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs]: I think he means giving the method a default implementation and call Tracker.OnGameEnd() inside this method

Comment (MaxIsJoe) [reply UnityProject/Assets/Scripts/MiniGames/MiniGameModule.cs]: I don't like the idea of turning this into a virtual void but since this PR has been taking forever.. eh, whatever. done.

**Review by corp-0 — APPROVED** [review]
